zkvm/src/types.rs

```diff
@@ -226,10 +226,28 @@ impl Into<Scalar> for ScalarWitness {
 }
 
 impl Predicate {
-    pub fn to_point(&self) -> CompressedRistretto {
-        match self {
-            Predicate::Opaque(point) => *point,
-            Predicate::Witness(witness) => witness.to_point(),
+    pub fn new(point: CompressedRistretto) -> Self {
+        Predicate {
+            point,
+            witness: None,
+        }
+    }
+
+    pub fn from_witness(witness: PredicateWitness) -> Result<Self, VMError> {
+        Ok(Predicate {
+            point: witness.to_point()?,
+            witness: Some(witness),
+        })
+    }
+
+    pub fn point(&self) -> CompressedRistretto {
+        self.point
+    }
+
+    pub fn witness(&self) -> Option<PredicateWitness> {
```

Comment:
I'd return `&Option<PredicateWitness>` to avoid unnecessary cloning. Or, if that's more ergonomic, a `Option<&PredicateWitness>`. Then, the user can clone if they really need to, but if they encode the witness, they don't need to consume it.

Reply:
got it, that makes sense!